File: frontend/src/api/batchInstall.ts
import apiFetch from './client'

export interface BatchPrecheckResult {
  status: 'passed' | 'warning' | 'blocked'
  items: Array<{
    skill_id: string
    skill_name: string
    agent_id: string
    dep_check_status: string
    dep_check_detail: Array<Record<string, unknown>>
  }>
  total: number
  passed_count: number
  warning_count: number
  blocked_count: number
  summary: string
}

export interface BatchInstallQueue {
  id: string
  operation: string
  status: string
  total_items: number
  success_count: number
  fail_count: number
  warn_count: number
  precheck_status: string | null
  precheck_summary: string | null
  created_by: string
  created_at: string | null
  completed_at: string | null
}

export interface BatchInstallItem {
  id: string
  queue_id: string
  skill_id: string
  skill_name: string
  agent_id: string
  agent_name: string
  dep_check_status: string
  status: string
  error_message: string | null
  started_at: string | null
  completed_at: string | null
}

// 依赖预检
export async function batchPrecheck(skillIds: string[], agentIds: string[]): Promise<BatchPrecheckResult> {
  const resp = await apiFetch('/api/v1/batch-install/precheck', {
    method: 'POST',
    data: { skill_ids: skillIds, agent_ids: agentIds },
  })
  return resp.data
}

// 创建批量安装
export async function createBatchInstall(skillIds: string[], agentIds: string[], operation = 'install', createdBy = ''): Promise<BatchInstallQueue> {
  const resp = await apiFetch('/api/v1/batch-install', {
    method: 'POST',
    data: { skill_ids: skillIds, agent_ids: agentIds, operation, created_by: createdBy },
  })
  return resp.data
}

// 执行批量安装
export async function executeBatch(queueId: string): Promise<BatchInstallQueue> {
  const resp = await apiFetch(`/api/v1/batch-install/${queueId}/execute`, { method: 'POST' })
  return resp.data
}

// 查询队列列表
export async function listBatchQueues(status?: string, offset = 0, limit = 20): Promise<{ data: BatchInstallQueue[]; total: number }> {
  let url = `/api/v1/batch-install?offset=${offset}&limit=${limit}`
  if (status) url += `&status=${status}`
  const resp = await apiFetch(url, { method: 'GET' })
  return resp.data
}

// 获取队列详情
export async function getBatchQueue(queueId: string): Promise<BatchInstallQueue> {
  const resp = await apiFetch(`/api/v1/batch-install/${queueId}`, { method: 'GET' })
  return resp.data
}

// 获取队列项
export async function getBatchQueueItems(queueId: string, offset = 0, limit = 50): Promise<{ data: BatchInstallItem[]; total: number }> {
  const resp = await apiFetch(`/api/v1/batch-install/${queueId}/items?offset=${offset}&limit=${limit}`, { method: 'GET' })
  return resp.data
}

// 生成报告
export async function getBatchReport(queueId: string): Promise<Record<string, unknown>> {
  const resp = await apiFetch(`/api/v1/batch-install/${queueId}/report`, { method: 'GET' })
  return resp.data
}

File: backend/app/models/batch_install.py
"""
批量 Skill 分配与安装模型 - 依赖预检、安装队列、报告
"""
import uuid
from datetime import datetime, timezone

from sqlalchemy import Column, String, Text, DateTime, Integer
from sqlalchemy.sql import func

from app.db.session import Base


class BatchInstallQueue(Base):
    """批量安装队列"""
    __tablename__ = "batch_install_queues"

    id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    operation = Column(String(20), nullable=False)  # install | uninstall | bind | unbind
    status = Column(String(20), default="pending")  # pending | prechecking | running | completed | failed | cancelled

    # 统计
    total_items = Column(Integer, default=0)
    success_count = Column(Integer, default=0)
    fail_count = Column(Integer, default=0)
    warn_count = Column(Integer, default=0)

    # 依赖预检
    precheck_status = Column(String(20))  # passed | warning | blocked
    precheck_summary = Column(Text)  # JSON

    # 来源
    created_by = Column(String(36))
    created_at = Column(DateTime(timezone=True), server_default=func.now(), default=lambda: datetime.now(timezone.utc))
    updated_at = Column(DateTime(timezone=True), onupdate=func.now(), default=lambda: datetime.now(timezone.utc))
    completed_at = Column(DateTime(timezone=True))


class BatchInstallItem(Base):
    """批量安装项（每个 Skill-Agent 对）"""
    __tablename__ = "batch_install_items"

    id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    queue_id = Column(String(36), nullable=False, index=True)
    skill_id = Column(String(36), nullable=False)
    skill_name = Column(String(100))
    agent_id = Column(String(36), nullable=False)
    agent_name = Column(String(100))

    # 依赖预检结果
    dep_check_status = Column(String(20))  # passed | warning | blocked
    dep_check_detail = Column(Text)  # JSON: 每个依赖的检查详情

    # 安装状态
    status = Column(String(20), default="pending")  # pending | running | success | failed | skipped
    error_message = Column(Text)
    started_at = Column(DateTime(timezone=True))
    completed_at = Column(DateTime(timezone=True))

    created_at = Column(DateTime(timezone=True), server_default=func.now(), default=lambda: datetime.now(timezone.utc))
